### PR TITLE
Dockerizing xarm-grpc

### LINKED ISSUE #33: Dockerfile for xarm-grpc-service
The Dockerfile for xarm-grpc-service should come in handy to validate the build as well as cleaning up the build process.

### LINKED ISSUE #36: Build test in the GitHub Action
Due to the complex dependency, we need to validate the build on the `main` branch always succeeds.
Add a workflow for build test including the generation of grpc sources from the proto file.

### PR DESCRIPTION
**Pull request by S-Eibl**
Add a Dockerfile to the xarm-grpc project. Loading cmake, grpc and the project on a Ubuntu 18.04.lts

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -0,0 +1,50 @@
+FROM ubuntu:18.04 AS base
```
Comment (iomz) [inline Dockerfile:1]: Why 18.04 instead of the latest LTS 20.04?

File: Dockerfile [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+    pushd cmake/build; \
+    cmake -DgRPC_INSTALL=ON \
+          -DgRPC_BUILD_TESTS=OFF \
+          -DCMAKE_INSTALL_PREFIX=$MY_INSTALL_DIR \
+          ../..; \
+    make; \
+    make install; \
+    popd;
+
+FROM extend-grpc AS xarm-grpc
```
Comment (iomz) [inline Dockerfile:35]: Again, meaningless staging.

File: Dockerfile [hunk 3]
```diff
@@ -0,0 +1,50 @@
+FROM ubuntu:18.04 AS base
+
+ENV CMAKE_VERSION=3.22.0
```
Comment (iomz) [inline Dockerfile:3]: Do we really need v3.22.0 cmake?

**Review by iomz — CHANGES_REQUESTED** [review]
Please respect the commit log rules: https://github.com/Interactions-HSG/xarm-grpc/blob/main/CONTRIBUTING.md#commits

The resulting image from this Dockerfile is 3.28GB (!). Some parts can be optimized to reduce the final image. I'm pushing the changes.

Instead of using cmake, building xarm-grpc with Bazel would make things more straightforward(which I'm not familiar with...)
https://github.com/grpc/grpc/blob/master/BUILDING.md#building-with-bazel-recommended

**Comment by iomz** [comment]
[792b745](https://github.com/Interactions-HSG/xarm-grpc/pull/53/commits/792b74591e66f2ba9a80f8d47b95ee192e2f07da) resolves #33 and #36

This is ugly. Further improvements are desired...

File: .github/workflows/docker-build.yml [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+          submodules: true
+      - name: Set up QEMU
+        uses: docker/setup-qemu-action@v1
+      - name: Set up Docker Buildx
+        uses: docker/setup-buildx-action@v1
+      - name: Login to DockerHub
+        uses: docker/login-action@v1
+        with:
+          username: ${{ secrets.DOCKERHUB_USERNAME }}
+          password: ${{ secrets.DOCKERHUB_TOKEN }}
```
Comment (iomz) [inline .github/workflows/docker-build.yml:24]: Currently using my credential

File: .github/workflows/docker-build.yml [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+      - name: Build and export to Docker
+        uses: docker/build-push-action@v2
+        with:
+          context: .
+          load: true
```
Comment (iomz) [inline .github/workflows/docker-build.yml:29]: Not pushing to Docker Hub registry

File: Dockerfile [hunk 6]
```diff
@@ -0,0 +1,39 @@
+# syntax=docker/dockerfile:1
+ARG GCC_VERSION=11.2.0
+ARG CMAKE_VERSION=3.22.0
+ARG GRPC_VERSION=1.43.0
+
+FROM iomz/cmake:$CMAKE_VERSION AS cmake
+FROM iomz/grpc-protoc:$GRPC_VERSION AS grpc
```
Comment (iomz) [inline Dockerfile:7]: Since the gRPC image takes insanely long to build, I cheated here. I will come back to this with #54.

File: Dockerfile [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+
+# this stage builds xarm-grpc
+FROM gcc:$GCC_VERSION
```
Comment (iomz) [inline Dockerfile:10]: This is also nasty. Ideally, I want to just cross-compile the binary for alpine...

**Review by iomz — APPROVED** [review]
@S-Eibl Sorry for the destructive changes. What do you think?

**Comment by iomz** [comment]
@S-Eibl I need your comment to merge this PR.

**Comment by iomz** [comment]
I didn't ask you to merge silently...

**Comment by S-Eibl** [comment]
Oh sorry, should I revert that?

**Comment by iomz** [comment]
> Oh sorry, should I revert that?

Nope, it would've been just nice if you "responded" to the comments/review/code, e.g., "Looks good to me", "I don't understand this: ...", "This part needs to be revised in the following manner in the future", etc. The reviewing is a process of problem resolution, and without it, it is hard to track the decision making process of the developers (us).

**Pull request merged**